### PR TITLE
Introduce FXGraphExtractor into torch.onnx.dynamo_export

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
The current API architecture can be seen as 3 independent exporters as shown below. The public API `dynamo_export()` defaults to one of the 3 variants and the other 2 must be used by instantiating private classes: image

This PR refactors the API in a way that `dynamo_export` is the only way to use the ONNX exporter. It defaults to a FX tracer based on ``torch.export``, but an internal-only idiom allows switching the FX tracer (aka `FXGraphExtractor` interface), as shown below:

image

Summary of changes:

* Unifies all exporter variants under a single `dynamo_export` API
  * `ResolvedExportOptions` was expanded to allow `fx_tracer: FXGraphExtractor` to be specified, selecting which FX graph extractor to use, according to the design proposal
  * As a consequence, `torch.onnx._internal.exporter.Exporter` does not have to *internally* specialize for each type of FX API that the exporter might be used. This leads to a single `Exporter` with many `FX graph extractors`
  * Before in red, after in green: image
* Input processing was moved from `Exporter` subclasses to `FXGraphExtractor` subclasses, where they are actually consumed
  * `Exporter` is a [data]class that holds export options, model and input data in a single cohesive object. Specializing it means create different exporters instead of having one exporter capable of exporting models through different options.
  * `Exporter` doesn't consume the `model_args` that caused it to specialize
* Improved the circular dependency story.
  * https://github.com/pytorch/pytorch/pull/99070 moves `import torch.onnx` to after all dynamo subcomponents, preventing `torch.onnx` to have circular depemndencies when `torch.XXXX` is imported during initialization
  * There are other points we need to improve in subsequent PRs. APIs are organized in a way that it is easy to "import too much"
* Refactored `decomposition_table` as an internal-only `ResolvedExportOptions` property.
  * Similar to input processing, this helper is not actually consumed at tyhe `Exporter` layer. This PR moves it to the layer in which it is used
* Demoted `Exporter.model_signature` to a simple standalone helper
  * There is no need to have this as a exporter method; this is a standard `inpect.signature` usage without any state

Possible next steps are:
* Decouple `passes` and `dispatching` from the cluttered `export_fx_to_onnx`
* Further integration with http://github.com/pytorch/pytorch/pull/98421/ into `FXGraphExtractor` public API + helper for unit testing
  * Some passes are changing input processing, which are not captured by the proposed input adapter

** COPILOT SUMMARY**

### Summary
📝🚀🔧

This pull request refactors the ONNX exporter code to use the FX graph representation and the new `io_adapter` module for input and output adaptation. It also adds support for custom export options and flattening HuggingFace model outputs in the ONNX test framework. It updates the ONNX dynamo exporter API tests and adds a new module `torch/onnx/_internal/fx/dynamo_graph_extractor.py` for exporting FX models to ONNX with dynamo support. It fixes some type annotations, imports, and formatting issues in the ONNX exporter code.

> _The ONNX exporter got a new look_
> _With FX graph and dynamo hook_
> _It uses `io_adapter`_
> _And custom options matter_
> _For HuggingFace models and `model_signature` book_

### Walkthrough
*  Move the `fx` submodule from `torch/onnx/_internal` to `torch/onnx/_internal/fx`, and rename some of its modules ( [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-c8fa56eefd7f98fb4f9739d57df57f02ede77e28528133736010a6d06651ebcbL21-R26), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L25-R26), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-3eef404cb9d85216c050be153c33255ebce1170a77d8b9b17be79bcfb238c9c4L5-R15), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-4da17ba9e1a187bfacb65a70d6ff15f6c2a60480be8e20fc452d8984a279cd0aL3-R30))
*  Add a new module `torch/onnx/_internal/fx/dynamo_graph_extractor.py` that defines a `DynamoExport` class for generating FX graphs using the `torch._dynamo.export` API ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-078d7b8d0e4050e650fc3c15dc97a0564852191ac7b7bdc069d0b3959c5ee39aR1-R77))
*  Add a new module `torch/onnx/_internal/fx/io_adapter.py` that defines the input and output adapter classes and steps for the ONNX exporter, and a helper function to wrap models with output adapters ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L159-R192), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-4da17ba9e1a187bfacb65a70d6ff15f6c2a60480be8e20fc452d8984a279cd0aL3-R30), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-4da17ba9e1a187bfacb65a70d6ff15f6c2a60480be8e20fc452d8984a279cd0aR72-R176), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-4da17ba9e1a187bfacb65a70d6ff15f6c2a60480be8e20fc452d8984a279cd0aL237-R478))
*  Update the `ResolvedExportOptions` class in `torch/onnx/_internal/exporter.py` to inherit from the `ExportOptions` class, and to set the `fx_tracer` and `decomposition_table` attributes based on the `dynamo_graph_extractor` and `function_dispatcher` modules ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L81-R99), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862R117-R126))
*  Update the `Exporter` class in `torch/onnx/_internal/exporter.py` to remove the `export` method and add a new abstract `generate_fx` method, and to use the `fx_tracer` attribute to generate and export the FX graph ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L413-R475), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L422-R486))
*  Update the `FXSymbolicTraceExporter` class in `torch/onnx/_internal/fx/fx_symbolic_graph_extractor.py` to be renamed to `FXSymbolicTracer`, and to inherit from `exporter.FXGraphExtractor` and implement the `generate_fx` method ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-3eef404cb9d85216c050be153c33255ebce1170a77d8b9b17be79bcfb238c9c4L128-R175), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-3eef404cb9d85216c050be153c33255ebce1170a77d8b9b17be79bcfb238c9c4L157-R219))
*  Update the `export_fx_to_onnx` method of the `FXSymbolicTracer` class to be renamed to `_export_fx_to_onnx`, and to be moved to the `exporter.FXGraphExtractor` class ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-3eef404cb9d85216c050be153c33255ebce1170a77d8b9b17be79bcfb238c9c4L193-R234))
*  Update the `dynamo_export` function in `torch/onnx/_internal/exporter.py` to accept and return `ResolvedExportOptions` and `Exporter` objects, respectively ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L536-R606))
*  Update the `run_test_with_fx_to_onnx_exporter_and_onnx_runtime` function in `test/onnx/onnx_test_common.py` to add a new parameter `export_options` for passing custom export options to the `torch.onnx.dynamo_export` function ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-1b38383dc1a0228a835d83bb7c4ba2d0c1bcd41297be5c6336572c525846166eR176), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-1b38383dc1a0228a835d83bb7c4ba2d0c1bcd41297be5c6336572c525846166eL216-R222))
*  Update the `test_log_sigmoid` and `_test_large_scale_exporter` tests in `test/onnx/test_fx_to_onnx_with_onnxruntime.py` to use the updated `run_test_with_fx_to_onnx_exporter_and_onnx_runtime` function and the `torch.onnx.dynamo_export` function ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-c8fa56eefd7f98fb4f9739d57df57f02ede77e28528133736010a6d06651ebcbL297-R301), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-c8fa56eefd7f98fb4f9739d57df57f02ede77e28528133736010a6d06651ebcbL682-R686), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-c8fa56eefd7f98fb4f9739d57df57f02ede77e28528133736010a6d06651ebcbL696-R716), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-c8fa56eefd7f98fb4f9739d57df57f02ede77e28528133736010a6d06651ebcbL721-R730))
*  Update the `test_raise_on_invalid_save_argument_type` test in `test/onnx/dynamo/test_exporter_api.py` to use the `io_adapter.InputAdapter` and `io_adapter.OutputAdapter` classes instead of the `exporter.InputAdapter` and `exporter.OutputAdapter` classes ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-4545f0c15c73ebe90a875e9bee6c5ca4b6b92fb1ed0ec5560d1568e0f6339d02L139-R139))
*  Move the `model_signature` property from the `Exporter` class in `torch/onnx/_internal/exporter.py` to a standalone function in `torch/onnx/utils.py`, and update the references to it ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L432-R505), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-3eef404cb9d85216c050be153c33255ebce1170a77d8b9b17be79bcfb238c9c4L157-R219), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-849a5778e2dcf7f36587967273cee0bf20642e35bf4c79405111ea3417c3fb3cL54-R75))
*  Move the `UnsatisfiedDependencyError` class from the `Exporter` class in `torch/onnx/_internal/exporter.py` to the top level of the module, and update the references to it ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L442-R512))
*  Rename the `_create_onnx_friendly_decomposition_table` function and the `_ONNX_FRIENDLY_DECOMPOSITION_TABLE` dictionary in `torch/onnx/_internal/fx/function_dispatcher.py` to `_create_default_onnx_decomposition_table` and `_DEFAULT_ONNX_EXPORTER_DECOMPOSITION_TABLE`, respectively, and update the references to them ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL213-R219), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL231-R239))
*  Update the imports in `torch/onnx/_internal/fx/function_dispatcher.py` to use the `torch._ops` and `torch._decomp` modules instead of the `torch.ops` and `torch.decomp` modules, and to use aliases for accessing the `onnxscript.function_libs.torch_aten.ops` and `torch._ops` modules ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL11-R16), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL35-R156), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL160-R166), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL173-R182), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL189-R194), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL201-R204), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-549890bc593f917c4e62c4c43077340e4774c0abdf31657ced8450fdfbed3b3eL231-R239))
*  Update the `ExportOutput` class in `torch/onnx/_internal/exporter.py` to use the `InputAdapter` and `OutputAdapter` classes from `io_adapter` instead of the ones defined in the same module ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L275-R199))
*  Update the type annotations in `torch/onnx/_internal/fx/serialization.py` and `torch/onnx/_internal/exporter.py` to fix some inconsistencies ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0c7a4333620a22a5c3e5315e30272b59fb7a11b393cb42f8255070bedeb02738L15-R15), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0c7a4333620a22a5c3e5315e30272b59fb7a11b393cb42f8255070bedeb02738L83-R83), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L11-R11), [link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862R18))
*  Remove an unused import of `inspect` from `torch/onnx/_internal/exporter.py` ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862L5))
*  Remove an unused import of `torch._dynamo` from `torch/onnx/_internal/fx/passes/shape_inference.py` ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-d38827b1f79525963c39e5c480240cd81f4edcaf8b3bd374a1c6ee2fdb28b334L7))
*  Add a comment to `torch/onnx/_internal/fx/passes/shape_inference.py` to explain why the import of `torch._dynamo` is done inside the `_run` method of the `ShapeInferenceWithFakeTensor` class ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-d38827b1f79525963c39e5c480240cd81f4edcaf8b3bd374a1c6ee2fdb28b334R32-R35))
*  Fix a typo in the docstring of the `_module_expansion_symbolic_trace` function in `torch/onnx/_internal/fx/fx_symbolic_graph_extractor.py` ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-3eef404cb9d85216c050be153c33255ebce1170a77d8b9b17be79bcfb238c9c4L96-R98))
*  Add an empty line to `torch/onnx/__init__.py` for formatting purposes ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-c3c8c09b65c1235ca4494633c6a0aab2761a11a7653ddaf9f874bbcd91e15553R12))
*  Delete the `torch/onnx/_internal/fx/__init__.py` file ([link](https://github.com/pytorch/pytorch/pull/99940/files?diff=unified&w=0#diff-a39fa3741f027bb9717388fc922d1e846fbd43d44f2c5fbee4e8d2188a7edb85))

Fixes #ISSUE_NUMBER

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:96]: docstring not accurate, more like 

`Dictionary with ATen/Prims operators and their decompositions. Exporter runs decomposition after getting the FX graph from FXGraphExtractor. If not provided, the default decomposition table is based on the full PyTorch decomposition table, excluding Prims decomposition, and any ATen operator decomposition if the ATen operator has a matching OnnxFunction in torch_lib.`

And based on the level of internal details, not sure why it needs to be exposed in this PR. Unless it is needed urgently for Prims experiments. cc @justinchuby. Ideally a much more natural api emerges after the dispatcher refactor.

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:120]: well it is available on PyTorch onnx CI, but will take quite a while, if say possible, to be available on PyTorch non-onnx CI. I guess this comment is meant for the latter. But not sure why CI needs to be mentioned explicitly. It doesn't make sense for PyTorch non-onnx CI to have onnxscript if onnxscript is not a dependency of PyTorch.

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:569]: nit: abc style
```suggestion
        ...
```

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:453]: This feels weird to be here. This class is `FXGraphExtractor`, but this method finishes everything off with a onnx ModelProto. Feels this method belongs to `Exporter` (or `DynamoExporter(Exporter)` for the sake of import dep problems).

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [inline torch/onnx/_internal/exporter.py:98]: nit: prepend `_` if private?

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [inline torch/onnx/_internal/exporter.py:120]: We can maybe also delay the table evaluation?

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [inline torch/onnx/_internal/exporter.py:460]: Possible to move out?

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [inline torch/onnx/_internal/exporter.py:584]: Maybe just make `ResolvedExportOptions()` idempotent to combine both branches?

File: torch/onnx/_internal/fx/function_dispatcher.py
Comment (justinchuby) [inline torch/onnx/_internal/fx/function_dispatcher.py:234]:
```suggestion
DEFAULT_ONNX_EXPORTER_DECOMPOSITION_TABLE: Dict[
```

I would make public if it is used in other modules

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:453]: To be honest, there is no right place on the repo for `_export_fx_to_onnx`. Next round of design can split it in several methods in their correct places.

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:96]: This is not publicly exposed; `ResolvedExportOptions` is internal-only. This serves as an internal extension point and it is a natural Dynamo concept

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:120]: There is a limitation on maximum line length, so we can't be too verbose.
It is assumed who reads this comment knows they are dealing with `torch.onnx`-only namespace already, right?

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:460]: `_export_fx_to_onnx` weill be entirely remove in next design stages, there is no point in trying to fix it in this PR

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:569]: Pytorch uses Google style, which uses `pass`?

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:584]: Not sure what you meant

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:460]: SG! I suggest a simple comment so we don't get confused in the interim.

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:569]: Looks like `...` is what python uses nowadays (I didn't know either): https://docs.python.org/3/library/abc.html#abc.abstractmethod

Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:569]: I personally suggest `raise NotImplementedError()` because otherwise subclasses can still call `super().xxx()` and silently do nothing

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:584]: Instead of casing here, move this logic to the initializer of `ResolvedExportOptions` so we can collapse the isinstance calls on the call site to simply

```
self.options = ResolvedExportOptions(options)
```

always.

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:98]: Keptit without `_` because it is used outside `exporter.py` scope, although inside `_internal` folder. This is the same criteria we used to change `_ResolvedExportOptions` to `ResolvedExportOptions`. Both are defined at `exporter.py` and used in different files.

Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:98]: SG! In that case I would just refine the comment to clarify it is not a private field to the class.

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:120]: `ResolvedExportOptions` is only evaluated when an export has started. This is as delayed as it can be.
`decomposition_table` is not evaluated when we do `import torch` or `import torch.onnx`, etc.

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:98]: The comment above states "# Private only attributes" but the fields do not have private names. This creates confusion for ourselves.

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [reply torch/onnx/_internal/exporter.py]: Sorry I meant the DEAFAULT_xxx table so as to avoid requiring onnxscript when importing these lines? Just a thought. May not be a good idea.

File: torch/onnx/_internal/exporter.py
Comment (justinchuby) [reply torch/onnx/_internal/exporter.py:98]: Perhaps just change that to `# Internal attributes` or `# Exporter internal attributes` etc.? Using `internal` removes that ambiguity because "private" tends to mean a very specify scope

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py]: Oh, I did that the first time, but then I changed because:
* beartype didn't like `ResolvedExportOptions.__init__(self, options: Optional[Union[ExportOptions, ResolvedExportOptions]])`, so I had to change it to `Optional[Union[ExportOptions, "ResolvedExportOptions"]]` to prevent `beartype.roar.BeartypePep563Exception: @beartyped torch.onnx._internal.exporter.ResolvedExportOptions.__init__() parameter "options" PEP 563-postponed type hint 'Optional[Union[ExportOptions, ResolvedExportOptions]]' syntactically invalid (i.e., "name 'ResolvedExportOptions' is not defined") `
* Python doesn't have copy constructor, so I would have to manually assign all `options`'s properties to `self`, making yet another redundancy assignment. Everytime we add a new option, we would have to change in "fake copy ctor".

I wonder why you didn't say anything when you approved https://github.com/pytorch/pytorch/pull/97920/files#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862R218-R222 which introduced this branching on `Exporter` instead of `ResolvedExportOptions`. My only change was to switch the order of comparison, but the branching was there already :) 

Nonetheless, I did this change as suggested.

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:96]: Using docstring from https://github.com/thiagocrepaldi/pytorch/blob/c8bf6f3c699bfd9833e9b9b65a1c39878f1ad440/torch/_dynamo/eval_frame.py#L705

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:98]: Thank you for your suggestion, but this is a won't fix. Too minor

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py]: Yeah, that is something I also feel needs another round of refactoring. We shouldn't have code being called on global scope as `decomp.py` does. 

This can be tackled along with the `passes` redesign, as decomposition is done as a pass today

Comment (BowenBao) [reply torch/onnx/_internal/exporter.py]: My point is the real problem is already mentioned as `[onnx]script dep` up there. I don't understand what is special about "CI" that it needs to be brought up.

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py]: This is similar to the docstring for `backend` of `DynamoOptimize` that we discussed earlier. Both need code rework to match the same concept of its original fx/dynamo counterpart, which is what stated in the docstring. I won't block on this, will revisit in dispatcher theme.

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py:462]: it is the "fx passes" + "conversion to onnxscript" part. Again as discussed my hunch is they should be the general part of exporter. I'm just surprised that it is moved from `DynamoExporter` to `Extractor`. So now all the work e2e is completed in `Extractor`, that feels like just a rename of the previous `FXExporter` though. Anyway won't block on this too, as you mentioned there will be follow up designs.

**Review by BowenBao — APPROVED** [review]
LGTM, with separate follow-ups for dispatcher(decomposition_table) and _export_fx_to_onnx next.

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:462]: `class Exporter` as proposed by Aaron is just a thread-safety container to keep user model+data+export options together so that we could run exporter in parallel (in theory). It was never meant to have hold code. The fact we did specialize and put a lot of helpers on it is the actual infraction we need to address.

`_export_fx_to_onnx` does too much (passes, graph building and protobuf serialization) and not even `Exporter` is the right place to it. In next steps, we will be able to break all three tasks in 3 different blocks, and neither `Exporter` nor `FXGraphExtractor` will need to hold any of it.

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py]: sorry, dude, the rephrasing somehow was `git add`ed on my last commit.

The comment will now read:
```python
import torch.onnx._internal.fx.dynamo_graph_extractor as dynamo_graph_extractor  # TODO: Prevent circular dep
from torch.onnx._internal.fx import (  # TODO: PyTorch does not take dep on onnxscript outside torch.onnx context
```

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py]: I don't get why `"""A dictionary that maps operators to their decomposition functions."""` doesn't match the same concept.

I understand we are not giving a full description on how it is used, but it is still a truthful explanation;  we do build a decomposition_table (with whatever our registry can support) and pass it to FX APIs o consume it, as they document it they should be.

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py:461]: I agree these are 3 separate tasks. In my understanding the `Exporter`, instead of `FXGraphExtractor`, should drive these components.

File: torch/onnx/_internal/exporter.py
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py]: Fair enough. I felt we are adding it before designing the purpose and the expected outcome under exporter context, which is what dispatcher design going to do anyway.

Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py]: got your point.
not all changes here are final because the rest of the code is not final.
next redesigns may refactor some of the things done on previous refactoring until everything make sense :)

File: torch/onnx/_internal/exporter.py
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:461]: Yeah, agreed. `Exporter` must drive the 3 classes that are to be born.
`_export_fx_to_onnx` will be gone and not inside fx_tracer very soon

File: torch/_dynamo/eval_frame.py
Comment (BowenBao) [inline torch/_dynamo/eval_frame.py:82]: yikes this needs additional approval now lol. I wonder if it makes sense to include whole 'onnx/_internal/fx` directory or not 🤔. Couldn't tell since really have little idea what this is doing.

Comment (thiagocrepaldi) [reply torch/_dynamo/eval_frame.py:82]: @kit1980 or @malfet maybe could also help.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**
